File: core/math/math_engine.js
import { vec3, mat4, quat } from "gl-matrix";

export class MathEngine {

    // --- Vector Operations ---

    static add(a, b) {
        const out = vec3.create();
        vec3.add(out, a, b);
        return out;
    }

    static sub(a, b) {
        const out = vec3.create();
        vec3.sub(out, a, b);
        return out;
    }

    static dot(a, b) {
        return vec3.dot(a, b);
    }

    static cross(a, b) {
        const out = vec3.create();
        vec3.cross(out, a, b);
        return out;
    }

    static normalize(v) {
        const out = vec3.create();
        vec3.normalize(out, v);
        return out;
    }

    // --- Matrix Operations ---

    static identity() {
        return mat4.create();
    }

    static translate(v) {
        const out = mat4.create();
        mat4.translate(out, out, v);
        return out;
    }

    static scale(v) {
        const out = mat4.create();
        mat4.scale(out, out, v);
        return out;
    }

    static rotate(radians, axis) {
        const out = mat4.create();
        mat4.rotate(out, out, radians, axis);
        return out;
    }

    // --- Quaternion Operations ---

    static fromEuler(eulerRadians) {
        const q = quat.create();
        quat.fromEuler(q,
            eulerRadians[0] * (180 / Math.PI),
            eulerRadians[1] * (180 / Math.PI),
            eulerRadians[2] * (180 / Math.PI)
        );
        return q;
    }

    static toEuler(q) {
        const out = vec3.create();
        const [x, y, z, w] = q;

        // Roll (x-axis)
        const sinr = 2 * (w * x + y * z);
        const cosr = 1 - 2 * (x * x + y * y);
        out[0] = Math.atan2(sinr, cosr);

        // Pitch (y-axis)
        const sinp = 2 * (w * y - z * x);
        out[1] = Math.abs(sinp) >= 1
            ? Math.sign(sinp) * Math.PI / 2
            : Math.asin(sinp);

        // Yaw (z-axis)
        const siny = 2 * (w * z + x * y);
        const cosy = 1 - 2 * (y * y + z * z);
        out[2] = Math.atan2(siny, cosy);

        return out;
    }

    static slerp(a, b, t) {
        const out = quat.create();
        quat.slerp(out, a, b, t);
        return out;
    }

    // --- Interpolation ---

    static lerp(a, b, t) {
        return a + (b - a) * t;
    }

    static lerpVec3(a, b, t) {
        const out = vec3.create();
        vec3.lerp(out, a, b, t);
        return out;
    }
}

File: native/ghost_core/src/MathEngine.cpp
#include "MathEngine.h"

namespace ghost_core {

    // --- Vector Operations ---

    glm::vec3 MathEngine::Add(const glm::vec3& a, const glm::vec3& b) {
        return a + b;
    }

    glm::vec3 MathEngine::Sub(const glm::vec3& a, const glm::vec3& b) {
        return a - b;
    }

    float MathEngine::Dot(const glm::vec3& a, const glm::vec3& b) {
        return glm::dot(a, b);
    }

    glm::vec3 MathEngine::Cross(const glm::vec3& a, const glm::vec3& b) {
        return glm::cross(a, b);
    }

    glm::vec3 MathEngine::Normalize(const glm::vec3& v) {
        return glm::normalize(v);
    }

    // --- Matrix Operations ---

    glm::mat4 MathEngine::Identity() {
        return glm::mat4(1.0f);
    }

    glm::mat4 MathEngine::Translate(const glm::vec3& v) {
        return glm::translate(glm::mat4(1.0f), v);
    }

    glm::mat4 MathEngine::Scale(const glm::vec3& v) {
        return glm::scale(glm::mat4(1.0f), v);
    }

    glm::mat4 MathEngine::Rotate(float radians, const glm::vec3& axis) {
        return glm::rotate(glm::mat4(1.0f), radians, axis);
    }

    // --- Quaternion Operations ---

    glm::quat MathEngine::FromEuler(const glm::vec3& eulerRadians) {
        return glm::angleAxis(eulerRadians.z, glm::vec3(0.0f, 0.0f, 1.0f))
             * glm::angleAxis(eulerRadians.y, glm::vec3(0.0f, 1.0f, 0.0f))
             * glm::angleAxis(eulerRadians.x, glm::vec3(1.0f, 0.0f, 0.0f));
    }

    glm::vec3 MathEngine::ToEuler(const glm::quat& q) {
        return glm::eulerAngles(q);
    }

    glm::quat MathEngine::Slerp(const glm::quat& a, const glm::quat& b, float t) {
        return glm::slerp(a, b, t);
    }

    // --- Interpolation ---

    float MathEngine::Lerp(float a, float b, float t) {
        return a + t * (b - a);
    }

    glm::vec3 MathEngine::LerpVec3(const glm::vec3& a, const glm::vec3& b, float t) {
        return a + t * (b - a);
    }

} // namespace ghost_core